### ISSUE TITLE
Using task instead of async, and offering query functions that fail hard instead of returning Result

### ISSUE THREAD

**Issue by aspnetde** [Zaid-Ajaj/Npgsql.FSharp#64]
I am using the library for accessing data for a Giraffe web application, and it's a pleasure to use so far! However, two things come to mind:

1. As Giraffe switched to https://github.com/rspeele/TaskBuilder.fs some time ago, it would be nice to also get tasks out of it.
2. In the context of a mobile app I would always prefer Result, however, for a web app I'd like to fail it hard and immediately -> throw exceptions.

So what I am doing right now is that:

```fs
module Sql =
    let throwOrReturn (result: Async<Result<'a, exn>>) =
        task {
            let! result = result |> Async.StartImmediateAsTask
            match result with
            | Ok result -> return result
            | Error exn -> return raise exn
        }

module FooRepository =
    let all =
        connect()
        |> Sql.query "..."
        ...
        |> Sql.throwOrReturn
```

I evaluate the result directly and either return the value or raise the exception. That's not perfect for many reasons, but the only workaround that comes to mind right now.

I also tried to make the adjustments as suggested myself, but as some things are private I cannot extend the existing library as I wish I could:

<img width="1038" alt="image" src="https://user-images.githubusercontent.com/1322584/89712153-49354500-d98f-11ea-8893-b9fb028fb138.png">

I see that switching from async to task would be a massive breaking change. Maybe it would be possible to add it as as second option. ~~As far as I can see, the parts within `async { ... }` would also work in `task { ... }`, so it wouldn't even require code duplication.~~ _Update: Forget about this, that's probably not going to work_.

The same goes for functions that do not return Result. Those could be used inside of the existing ones and a `try ... with` block.

What do you think?

**Comment by Zaid-Ajaj** [comment]
> it's a pleasure to use so far! 

Thanks, I am glad you are enjoying the library :smile: 

> What do you think?

I have a couple of solutions in mind but these tend to go against the spirit of the library: small and focused, no multiple options to do the same thing :wink: I re-implemented the library as of v3.0 and removed all task-returning functions so that people would do exactly what you are doing _when the need arises_ for example when using Giraffe . 

Is this `Sql.returnOrThrow` considered too much friction for you?

The solution I am thinking of right now is to expose a bunch of namespaces from this package:
 - `Npgsql.FSharp` does what it is doing right now `Async` + `Result`
 - `Npgsql.FSharp.Unsafe` exposes `Async` functions without Result
 - `Npgsql.FSharp.Tasks`  exposes `Task` + `Result` 
 - `Npgsql.FSharp.UnsafeTasks` exposes `Task` without the result

This solution removes the friction but as you can see, it might be a lot of work to build and to maintain. That's why that `Sql.returnOrThrow` might not be that bad after all :wink: 

Second problem is that I don't want to use tasks CE from a third-party package and would rather wait for the built-in task CE coming to F# 5.0 soon-ish

cc @TheAngryByrd any thoughts on the solution above?

**Comment by aspnetde** [comment]
> Second problem is that I don't want to use tasks CE from a third-party package and would rather wait for the built-in task CE coming to F# 5.0 soon-ish

For a library that makes sense.

> This solution removes the friction but as you can see, it might be a lot of work to build and to maintain.

I just took a shower and came up with a similar suggestion. But I agree, it's making things more complex while it remains unclear how many people would even need it. (Although a lot of stuff could be shared between those different flavors.)

> Is this `Sql.returnOrThrow` considered too much friction for you?

It's okay. I still feel not familiar enough yet with both `async` and `task`, so I can only hope that it does not cause any surprising and unwanted side effects.

Maybe making less (or no) things private would enable anyone to extend and tweak it as they wish?

**Comment by Zaid-Ajaj** [comment]
> Maybe making less (or no) things private would enable anyone to extend and tweak it as they wish?

Yeah but that means the API will expose functions in `Sql` module which many users will never use. Not to mention how confusing it would be for beginners to have `connect`, `connectFromConfig` *and* a `getConnection` function (which they shouldn't use). I find the functions `connect` and `connectFromConfig` already confusing enough if you are just getting started. Trade-offs, trade-offs...

> I just took a shower and came up with a similar suggestion. But I agree, it's making things more complex while it remains unclear how many people would even need it. (Although a lot of stuff could be shared between those different flavors.)

I might actually do it for the next major v4 release, see the reaction and feedback from users before I revert everything back this version in v5 :joy: I mean opening a different namespace doesn't sound so bad (as long as you don't open more than one namespace....then multiple packages? arrggghhh)

**Comment by aspnetde** [comment]
> Trade-offs, trade-offs...

Those private things could live their life in a separate module, so the API of the `Sql` module would not be polluted ... ;)

**Comment by aspnetde** [comment]
> Is this Sql.returnOrThrow considered too much friction for you?

One thing it does is to obscure the stack trace. Exceptions raised will always appear with that function in the log, so there is a lot of guesswork necessary to track down the real cause of the error.

**Comment by TheAngryByrd** [comment]
> > Is this Sql.returnOrThrow considered too much friction for you?
> 
> One thing it does is to obscure the stack trace. Exceptions raised will always appear with that function in the log, so there is a lot of guesswork necessary to track down the real cause of the error.

It might be worth it for the library to consider using [ExceptionDispatchInfo.Capture](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?view=net-5.0#System_Runtime_ExceptionServices_ExceptionDispatchInfo_Capture_System_Exception_) as a way to handle exceptions so capturing the stacktrace remains accurate when a consumer throws it.

**Comment by Zaid-Ajaj** [comment]
> It might be worth it for the library to consider using ExceptionDispatchInfo.Capture as a way to handle exceptions so capturing the stacktrace remains accurate when a consumer throws it.

@TheAngryByrd Smells like a PR, doesn't it? 😄 Would really appreciate it if you send one since we are still using Result in the `Npgsql.FSharp` namespace

**Comment by Zaid-Ajaj** [comment]
Resolved as of v3.12.3
